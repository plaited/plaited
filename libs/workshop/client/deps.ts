// -- pixelmatch --
export { default as pixelmatch } from 'https://esm.sh/pixelmatch@5.3.0'

// -- axe-core--
// // @deno-types='npm:axe-core@4.6.3'
export { default as axe } from 'https://esm.sh/axe-core@4.6.3'
