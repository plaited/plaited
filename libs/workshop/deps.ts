// -- playwright --
export { type Page } from 'https://esm.sh/playwright@1.30.0'
export { type Expect } from 'https://esm.sh/@playwright/test@1.30.0'
