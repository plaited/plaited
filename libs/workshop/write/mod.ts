export * from './write-docker-compose.ts'
export * from './write-dockerfile.ts'
export * from './write-gitignore.ts'
export * from './write-package.ts'
export * from './write-playwright-config.ts'
export * from './write-workshop.ts'
export * from './write-spec.ts'
export * from './write-yarnrc.ts'
