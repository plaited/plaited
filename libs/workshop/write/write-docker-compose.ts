import { dockerCompose } from '../templates/mod.ts'
import { getStat } from '../../utils/mod.ts'

type WriteDockerCompose = (args: {
  pat: boolean
  path: string
  port: number
  project?: string
  protocol: 'http' | 'https'
}) => Promise<void>
export const writeDockerCompose: WriteDockerCompose = async ({
  pat,
  path,
  port,
  project,
}) => {
  const exist = await getStat(path)
  !exist && await Deno.writeTextFile(
    path,
    dockerCompose({
      pat,
      port,
      project,
    }),
  )
}
