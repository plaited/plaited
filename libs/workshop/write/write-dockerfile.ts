import { dockerFile } from '../templates/mod.ts'
import { getStat } from '../../utils/mod.ts'

export const writeDockerfile = async (path: string, pat: boolean) => {
  const exist = await getStat(path)
  !exist && await Deno.writeTextFileSync(
    path,
    dockerFile(pat),
  )
}
