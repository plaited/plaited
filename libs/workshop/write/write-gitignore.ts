import { gitignore } from '../templates/mod.ts'
import { getStat } from '../../deno-utils/mod.ts'

export const writeGitignore = async (path: string) => {
  const exist = await getStat(path)
  !exist && await Deno.writeTextFileSync(
    path,
    gitignore,
  )
}
