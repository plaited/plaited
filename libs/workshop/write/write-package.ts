import { packageJson } from '../templates/mod.ts'
import { getStat } from '../../deno-utils/mod.ts'

export const writePackage = async (path: string) => {
  const exist = await getStat(path)
  !exist && await Deno.writeTextFileSync(
    path,
    packageJson,
  )
}
