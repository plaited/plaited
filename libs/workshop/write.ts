import { writeClient } from './write-client.ts'
import { Write } from './types.ts'
import { walk } from '../deps.ts'
import { writeTestRunner } from './write-test-runner.ts'
export const write: Write = async ({
  assets,
  exts,
  dev,
  importMap,
  workspace,
}) => {
  const { island, story, worker } = exts
  const workerExts = worker && Array.isArray(worker)
    ? worker
    : worker
    ? [worker]
    : []
  const islandExts = Array.isArray(island) ? island : [island]
  const storyExts = Array.isArray(story) ? story : [story]
  const combinedExts = [...islandExts, ...storyExts, ...workerExts]

  /** get paths and name for each island */
  const entryPoints: string[] = []
  for await (
    const entry of walk(workspace, {
      exts: combinedExts,
    })
  ) {
    const { path } = entry
    entryPoints.push(path)
  }

  const storyModules = entryPoints.filter((entry) =>
    storyExts.some((ext) => entry.endsWith(ext))
  )
  /** write client side code*/
  const clientEntries = await writeClient({
    dev,
    entryPoints,
    assets,
    importMap,
    workerExts,
  })

  const runnerEntry = await writeTestRunner({
    assets,
  })

  return { clientEntries, runnerEntry, storyModules }
}
