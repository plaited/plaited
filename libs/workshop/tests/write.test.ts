import {
  afterEach,
  assert,
  assertSnapshot,
  beforeEach,
  ConnInfo,
  describe,
  it,
} from '../../test-deps.ts'
import { write } from '../write.ts'
import { getStat } from '../get-stat.ts'
import { router } from '../../server/deps.ts'
import { resolve, toFileUrl } from '../../deps.ts'

const TEST_CONN_INFO: ConnInfo = {
  localAddr: {
    transport: 'tcp',
    hostname: 'test',
    port: 80,
  },
  remoteAddr: {
    transport: 'tcp',
    hostname: 'test',
    port: 80,
  },
}

const __dirname = new URL('.', import.meta.url).pathname
const assets = resolve(__dirname, './__mocks__/assets')
const root = resolve(__dirname, './__mocks__/root')
const playwright = resolve(__dirname, './__tmp__/write')
const importMap = toFileUrl(
  resolve(__dirname, '../../../.vscode/import-map.json'),
)

describe('Write', () => {
  beforeEach(async () => {
    await Deno.mkdir(assets, { recursive: true })
  })
  afterEach(async () => {
    const assetsExist = await getStat(assets)
    !assetsExist && await Deno.mkdir(assets, { recursive: true })
    await Deno.remove(assets, { recursive: true })
    const playwrightExist = await getStat(playwright)
    !playwrightExist && await Deno.mkdir(playwright, { recursive: true })
    await Deno.remove(playwright, { recursive: true })
  })
  it('build', async (t) => {
    const routes = await write({
      assets,
      root,
      importMap,
      playwright,
      dev: true,
      port: 3000,
      exts: {
        story: '.stories.ts',
        island: '.island.ts',
      },
    })
    for (const path in routes) {
      const route = router({
        [path]: routes[path],
      })
      const response = await route(
        new Request(`https://example.com${path}`),
        TEST_CONN_INFO,
      )
      const data = await response.arrayBuffer()
      const decoder = new TextDecoder()
      const text = decoder.decode(data)
      await assertSnapshot(t, text, { name: path })
    }
    const buttonSpec = await Deno.stat(
      `${playwright}/components/example-button.spec.js`,
    )
    const numberFieldSpec = await Deno.stat(
      `${playwright}/components/example-field.spec.js`,
    )
    assert(buttonSpec.isFile)
    assert(buttonSpec.size > 0)
    assert(numberFieldSpec.isFile)
    assert(numberFieldSpec.size > 0)
    const buttonStories = await Deno.stat(
      `${playwright}/.stories/example-button.stories.js`,
    )
    const numberFieldStories = await Deno.stat(
      `${playwright}/.stories/example-field.stories.js`,
    )
    assert(buttonStories.isFile)
    assert(buttonStories.size > 0)
    assert(numberFieldStories.isFile)
    assert(numberFieldStories.size > 0)
    const registry = await Deno.stat(
      `${assets}/.registries/example-field.island.js`,
    )
    assert(registry.isFile)
    assert(registry.size > 0)
  })
})
