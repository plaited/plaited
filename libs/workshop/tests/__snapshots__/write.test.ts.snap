export const snapshot = {};

snapshot[`/components-example-field--field-basic 1`] = `
\`<!DOCTYPE html><html lang="en"><head><meta charset="UTF-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Components Example Field(function(string) {
    return arrayReduce(words(deburr(string).replace(reApos, '')), callback, '');
  })</title><script type="module" async>import "/.registries/example-button.island.js"; import "/.registries/example-field.island.js";</script></head><body><example-field><template shadowroot="open"><style></style><input type='text' value="100" id="input" min="10" max="100" ref={ref} /></template></example-field><script type="text/javascript">const source = new EventSource('/livereload'); const reload = () =>location.reload(true); source.onmessage = reload; source.onerror = () =>(source.onopen = reload); console.log('[plaited] listening for file changes');</script></body></html>\`
`;

snapshot[`/components-example-field--field-basic.include 1`] = `\`<example-field><template shadowroot="open"><style></style><input type='text' value="100" id="input" min="10" max="100" ref={ref} /></template></example-field>\``;

snapshot[`/components-example-button--basic-button 1`] = `
\`<!DOCTYPE html><html lang="en"><head><meta charset="UTF-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Components Example Button(function(string) {
    return arrayReduce(words(deburr(string).replace(reApos, '')), callback, '');
  })</title><script type="module" async>import "/.registries/example-button.island.js"; import "/.registries/example-field.island.js";</script></head><body><plaited-fixture><template shadowroot="open"><style></style><button type="submit">basic</button></template></plaited-fixture><script type="text/javascript">const source = new EventSource('/livereload'); const reload = () =>location.reload(true); source.onmessage = reload; source.onerror = () =>(source.onopen = reload); console.log('[plaited] listening for file changes');</script></body></html>\`
`;

snapshot[`/components-example-button--basic-button.include 1`] = `'<plaited-fixture><template shadowroot="open"><style></style><button type="submit">basic</button></template></plaited-fixture>'`;
